### PR TITLE
[SPARK-18647][SQL] do not put provider in table properties for Hive serde table

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

In Spark 2.1, we make Hive serde tables case-preserving by putting the table metadata in table properties, like what we did for data source table. However, we should not put table provider, as it will break forward compatibility. e.g. if we create a Hive serde table with Spark 2.1, using `sql("create table test stored as parquet as select 1")`, we will fail to read it with Spark 2.0, as Spark 2.0 mistakenly treat it as data source table because there is a `provider` entry in table properties.

Logically Hive serde table's provider is always hive, we don't need to store it in table properties, this PR removes it.

## How was this patch tested?

manually test the forward compatibility issue.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @rxin @yhuai @mallman

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala [hunk 1]
```diff
@@ -232,17 +233,26 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
...
+    //  1. Put table metadata like table schema, partition columns, etc. in table properties.
     //  2. Check if this table is hive compatible.
     //    2.1  If it's not hive compatible, set location URI, schema, partition columns and bucket
     //         spec to empty and save table metadata to Hive.
     //    2.2  If it's hive compatible, set serde information in table metadata and try to save
     //         it to Hive. If it fails, treat it as not hive compatible and go back to 2.1
     val tableProperties = tableMetaToTableProps(table)
 
+    // put table provider and partition provider in table properties.
+    tableProperties.put(DATASOURCE_PROVIDER, provider)
```
Comment (mallman) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Why are we putting the provider name in the table properties here?

**Comment by mallman** [comment]
I built and tested this branch, and it resolves the issue I was having with reading Spark 2.1 tables in earlier versions of Spark. Thanks!

**Comment by rxin** [comment]
We need a test case here.

**Comment by gatorsmile** [comment]
We need to provide forward compatibility? That is pretty hard.

**Comment by rxin** [comment]
In a lot of environments people run multiple Spark versions side by side. That's always been a big strength of Spark.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala]: Previously we store the provider in the code path for both data source and hive serde tables. Now I move it to the data source table only code path.

**Comment by cloud-fan** [comment]
@rxin I'm afraid it's hard to write forward compatibility tests using unit test, we may need an external test infrastructure(python scripts) to do this.

**Comment by rxin** [comment]
We can have a test to check the table properties don't contain the entry, can't we?

**Comment by gatorsmile** [comment]
I see. Will be careful in the future to not break the forward compatibility.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala [hunk 2]
```diff
@@ -52,4 +54,18 @@ class HiveExternalCatalogSuite extends ExternalCatalogSuite {
...
+    val catalog = newBasicCatalog()
+    val hiveTable = CatalogTable(
+      identifier = TableIdentifier("hive_tbl", Some("db1")),
+      tableType = CatalogTableType.MANAGED,
+      storage = storageFormat,
+      schema = new StructType().add("col1", "int").add("col2", "string"),
+      provider = Some("hive"))
+    catalog.createTable(hiveTable, ignoreIfExists = false)
+
+    val rawTable = externalCatalog.client.getTable("db1", "hive_tbl")
```
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveExternalCatalogSuite.scala]: Could we also add one more check for another API `externalCatalog.getTable("db1", "hive_tbl")`? The provider should contain `DDLUtils.HIVE_PROVIDER`

**Comment by gatorsmile** [comment]
The [`alterTable`](https://github.com/cloud-fan/spark/blob/5ee6489cdd1c22a1071c4fb6c7e4c4af126c9d50/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala#L533) API in HiveExternalCatalog is still based on the provider field. We need a change too.

**Comment by cloud-fan** [comment]
@gatorsmile , `alterTable` don't need to change, the `provider` is not in table properties but still in the `CatalogTable.provider` field.

**Comment by gatorsmile** [comment]
@cloud-fan True. That is not the `oldTableDef` fetched from the metastore. 

LGTM pending test

**Comment by cloud-fan** [comment]
thanks for the review, merging to master/2.1!

**Pull request closed without merging**